### ISSUE TITLE
t/op/tie.t: blocker to strict-by-default

### ISSUE THREAD

**Issue by jkeenan** [atoomic/perl#239]
As of commit ac6b7ccf73 in the alpha-dev-02-strict branch on Aug 18 2020, the file cited is failing.  Here is output from `cd t; ./perl harness -v <file>; cd -`:
```
# t/op/tie.t

PROG: 

# standard behaviour, without any extra references
use Tie::Hash ;
tie %h, Tie::StdHash;
untie %h;
EXPECTED:

GOT:
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 4.
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 5.
Bareword "Tie::StdHash" not allowed while "strict subs" in use at - line 4.
Execution of - aborted due to compilation errors.
# Failed test 1 - at op/tie.t line 18
PROG: 
# (skip under miniperl if base.pm is not in lib/ yet)

# standard behaviour, without any extra references
use Tie::Hash ;
{package Tie::HashUntie;
 use base 'Tie::StdHash';
 sub UNTIE
  {
   warn "Untied\n";
  }
}
tie %h, Tie::HashUntie;
untie %h;
EXPECTED:
Untied
GOT:
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 12.
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 13.
Bareword "Tie::HashUntie" not allowed while "strict subs" in use at - line 12.
Execution of - aborted due to compilation errors.
# Failed test 2 - at op/tie.t line 18
PROG: 

# standard behaviour, with 1 extra reference
use Tie::Hash ;
$a = tie %h, Tie::StdHash;
untie %h;
EXPECTED:

GOT:
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 4.
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 5.
Bareword "Tie::StdHash" not allowed while "strict subs" in use at - line 4.
Execution of - aborted due to compilation errors.
# Failed test 3 - at op/tie.t line 18
PROG: 

# standard behaviour, with 1 extra reference via tied
use Tie::Hash ;
tie %h, Tie::StdHash;
$a = tied %h;
untie %h;
EXPECTED:

GOT:
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 4.
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 5.
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 6.
Bareword "Tie::StdHash" not allowed while "strict subs" in use at - line 4.
Execution of - aborted due to compilation errors.
# Failed test 4 - at op/tie.t line 18
PROG: 

# standard behaviour, with 1 extra reference which is destroyed
use Tie::Hash ;
$a = tie %h, Tie::StdHash;
$a = 0 ;
untie %h;
EXPECTED:

GOT:
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 4.
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 6.
Bareword "Tie::StdHash" not allowed while "strict subs" in use at - line 4.
Execution of - aborted due to compilation errors.
# Failed test 5 - at op/tie.t line 18
PROG: 

# standard behaviour, with 1 extra reference via tied which is destroyed
use Tie::Hash ;
tie %h, Tie::StdHash;
$a = tied %h;
$a = 0 ;
untie %h;
EXPECTED:

GOT:
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 4.
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 5.
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 7.
Bareword "Tie::StdHash" not allowed while "strict subs" in use at - line 4.
Execution of - aborted due to compilation errors.
# Failed test 6 - at op/tie.t line 18
PROG: 

# strict behaviour, without any extra references
use warnings 'untie';
use Tie::Hash ;
tie %h, Tie::StdHash;
untie %h;
EXPECTED:

GOT:
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 5.
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 6.
Bareword "Tie::StdHash" not allowed while "strict subs" in use at - line 5.
Execution of - aborted due to compilation errors.
# Failed test 7 - at op/tie.t line 18
PROG: 

# strict behaviour, with 1 extra references generating an error
use warnings 'untie';
use Tie::Hash ;
$a = tie %h, Tie::StdHash;
untie %h;
EXPECTED:
untie attempted while 1 inner references still exist at - line 6.
GOT:
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 5.
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 6.
Bareword "Tie::StdHash" not allowed while "strict subs" in use at - line 5.
Execution of - aborted due to compilation errors.
# Failed test 8 - at op/tie.t line 18
PROG: 

# strict behaviour, with 1 extra references via tied generating an error
use warnings 'untie';
use Tie::Hash ;
tie %h, Tie::StdHash;
$a = tied %h;
untie %h;
EXPECTED:
untie attempted while 1 inner references still exist at - line 7.
GOT:
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 5.
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 6.
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 7.
Bareword "Tie::StdHash" not allowed while "strict subs" in use at - line 5.
Execution of - aborted due to compilation errors.
# Failed test 9 - at op/tie.t line 18
PROG: 

# strict behaviour, with 1 extra references which are destroyed
use warnings 'untie';
use Tie::Hash ;
$a = tie %h, Tie::StdHash;
$a = 0 ;
untie %h;
EXPECTED:

GOT:
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 5.
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 7.
Bareword "Tie::StdHash" not allowed while "strict subs" in use at - line 5.
Execution of - aborted due to compilation errors.
# Failed test 10 - at op/tie.t line 18
PROG: 

# strict behaviour, with extra 1 references via tied which are destroyed
use warnings 'untie';
use Tie::Hash ;
tie %h, Tie::StdHash;
$a = tied %h;
$a = 0 ;
untie %h;
EXPECTED:

GOT:
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 5.
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 6.
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 8.
Bareword "Tie::StdHash" not allowed while "strict subs" in use at - line 5.
Execution of - aborted due to compilation errors.
# Failed test 11 - at op/tie.t line 18
PROG: 

# strict error behaviour, with 2 extra references
use warnings 'untie';
use Tie::Hash ;
$a = tie %h, Tie::StdHash;
$b = tied %h ;
untie %h;
EXPECTED:
untie attempted while 2 inner references still exist at - line 7.
GOT:
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 5.
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 6.
Global symbol "%h" requires explicit package name (did you forget to declare "my %h"?) at - line 7.
Bareword "Tie::StdHash" not allowed while "strict subs" in use at - line 5.
Execution of - aborted due to compilation errors.
# Failed test 12 - at op/tie.t line 18
PROG: 

# strict behaviour, check scope of strictness.
no warnings 'untie';
use Tie::Hash ;
$A = tie %H, Tie::StdHash;
$C = $B = tied %H ;
{
    use warnings 'untie';
    use Tie::Hash ;
    tie %h, Tie::StdHash;
    untie %h;
}
untie %H;
EXPECTED:

GOT:
Global symbol "$A" requires explicit package name (did you forget to declare "my $A"?) at - line 5.
Global symbol "%H" requires explicit package name (did you forget to declare "my %H"?) at - line 5.
Global symbol "$C" requires explicit package name (did you forget to declare "my $C"?) at - line 6.
Global symbol "$B" requires explicit package name (did you forget to declare "my $B"?) at - line 6.
Global symbol "%H" requires explicit package name (did you forget to declare "my %H"?) at - line 6.
BEGIN not safe after errors--compilation aborted at - line 8.
# Failed test 13 - at op/tie.t line 18
PROG: 

# Allowed IO self-ties
my $destroyed = 0;
sub Self3::TIEHANDLE { bless $_[1], $_[0] }
sub Self3::DESTROY   { $destroyed = 1; }
sub Self3::PRINT     { $printed = 1; }
{
    use Symbol 'geniosym';
    my $c = geniosym;
    tie *$c, 'Self3', $c;
    print $c 'Hello';
}
die "self-tied IO not PRINTed" unless $printed == 1;
die "self-tied IO not DESTROYed" unless $destroyed == 1;
EXPECTED:

GOT:
Global symbol "$printed" requires explicit package name (did you forget to declare "my $printed"?) at - line 6.
BEGIN not safe after errors--compilation aborted at - line 8.
# Failed test 17 - at op/tie.t line 18
PROG: 

# Interaction of tie and vec

my ($a, $b);
use Tie::Scalar;
tie $a,Tie::StdScalar or die;
vec($b,1,1)=1;
$a = $b;
vec($a,1,1)=0;
vec($b,1,1)=0;
die unless $a eq $b;
EXPECTED:

GOT:
Bareword "Tie::StdScalar" not allowed while "strict subs" in use at - line 6.
Execution of - aborted due to compilation errors.
# Failed test 19 - at op/tie.t line 18
PROG: 

# correct unlocalisation of tied hashes (patch #16431)
use Tie::Hash ;
tie %tied, Tie::StdHash;
{ local $hash{'foo'} } warn "plain hash bad unlocalize" if exists $hash{'foo'};
{ local $tied{'foo'} } warn "tied hash bad unlocalize" if exists $tied{'foo'};
{ local $ENV{'foo'}  } warn "%ENV bad unlocalize" if exists $ENV{'foo'};
EXPECTED:

GOT:
Global symbol "%tied" requires explicit package name (did you forget to declare "my %tied"?) at - line 4.
Global symbol "%hash" requires explicit package name (did you forget to declare "my %hash"?) at - line 5.
Global symbol "%hash" requires explicit package name (did you forget to declare "my %hash"?) at - line 5.
Global symbol "%tied" requires explicit package name (did you forget to declare "my %tied"?) at - line 6.
Global symbol "%tied" requires explicit package name (did you forget to declare "my %tied"?) at - line 6.
Bareword "Tie::StdHash" not allowed while "strict subs" in use at - line 4.
Execution of - aborted due to compilation errors.
# Failed test 20 - at op/tie.t line 18
PROG: 

# An attempt at lvalueable barewords broke this
tie FH, 'main';
EXPECTED:
Can't modify constant item in tie at - line 3, near "'main';"
Execution of - aborted due to compilation errors.
GOT:
Can't modify constant item in tie at - line 3, near "'main';"
Bareword "FH" not allowed while "strict subs" in use at - line 3.
Execution of - aborted due to compilation errors.
# Failed test 21 - at op/tie.t line 18
PROG: 
sub TIESCALAR { bless {} }
sub FETCH { my $x = 3.3; 1 if 0+$x; $x }
tie $h, "main";
print $h,"\n";
EXPECTED:
3.3
GOT:
Global symbol "$h" requires explicit package name (did you forget to declare "my $h"?) at - line 3.
Global symbol "$h" requires explicit package name (did you forget to declare "my $h"?) at - line 4.
Execution of - aborted due to compilation errors.
# Failed test 33 - at op/tie.t line 18
PROG: 
sub TIESCALAR { bless {} }
sub FETCH { shift()->{i} ++ }
tie $h, "main";
print $h.$h;
EXPECTED:
01
GOT:
Global symbol "$h" requires explicit package name (did you forget to declare "my $h"?) at - line 3.
Global symbol "$h" requires explicit package name (did you forget to declare "my $h"?) at - line 4.
Global symbol "$h" requires explicit package name (did you forget to declare "my $h"?) at - line 4.
Execution of - aborted due to compilation errors.
# Failed test 34 - at op/tie.t line 18
Variable "$IS_EBCDIC" is not imported at (eval 11) line 1, <DATA> line 1602.
# In SKIP code reason:
#  $IS_EBCDIC
Global symbol "$IS_EBCDIC" requires explicit package name (did you forget to declare "my $IS_EBCDIC"?) at (eval 11) line 1, <DATA> line 1602.
op/tie.t .. 
not ok 1
not ok 2
not ok 3
not ok 4
not ok 5
not ok 6
not ok 7
not ok 8
not ok 9
not ok 10
not ok 11
not ok 12
not ok 13
ok 14
ok 15
ok 16
not ok 17
# PROG: 
# 
# my $destroyed = 0;
# sub Self3::TIEHANDLE { bless $_[1], $_[0] }
# sub Self3::DESTROY   { $destroyed = 1; }
# sub Self3::PRINT     { $printed = 1; }
# {
#     use Symbol 'geniosym';
#     my $c = geniosym;
#     tie *$c, 'Self3', \*$c;
#     print $c 'Hello';
# }
# die "IO tied to TEMP glob not PRINTed" unless $printed == 1;
# die "IO tied to TEMP glob not DESTROYed" unless $destroyed == 1;
# EXPECTED:
# 
# GOT:
# Global symbol "$printed" requires explicit package name (did you forget to declare "my $printed"?) at - line 5.
# BEGIN not safe after errors--compilation aborted at - line 7.
not ok 18 # TODO IO "self-tie" via TEMP glob
# Failed test 18 - at op/tie.t line 18
not ok 19
not ok 20
not ok 21
ok 22
...
ok 32
not ok 33
not ok 34
Dubious, test returned 255 (wstat 65280, 0xff00)
Failed 19/34 subtests 

Test Summary Report
-------------------
op/tie.t (Wstat: 65280 Tests: 34 Failed: 19)
  Failed tests:  1-13, 17, 19-21, 33-34
  Non-zero exit status: 255
  Parse errors: No plan found in TAP output
Files=1, Tests=34,  0 wallclock secs ( 0.01 usr  0.01 sys +  0.19 cusr  0.05 csys =  0.26 CPU)
Result: FAIL
```
Probably encountered a rabbit hole once I started to properly scope those variables.

@atoomic @toddr 

Thank you very much.
Jim Keenan

**Comment by jkeenan** [comment]
This should be fixed with a88fff3981, but would benefit from review.  @atoomic

**Comment by atoomic** [comment]
LGTM we can close this case
